Cache: create docs

zucchini-nlp opened the pull request:
# What does this PR do?

This PR creates docs for cache classes and adds example usage in each classes docstring. Given that we often get issue that cache doesn't work, this docs page should help us to decrease our notifications. Partially addresses #31795, but I'll make another PR for that issue changing docs of each model

I made a quick docs, so let me know if anything was missed or needs a modification 
cc @gante @ArthurZucker

zucchini-nlp committed:
draft

zucchini-nlp committed:
Merge remote-tracking branch 'upstream/main' into cache_docs

zucchini-nlp committed:
updates

zucchini-nlp committed:
works?

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_32150). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

zucchini-nlp committed:
try adding python example in hidden section

zucchini-nlp committed:
another try

zucchini-nlp committed:
hwo do i render python

zucchini-nlp committed:
format as html code?

zucchini-nlp committed:
Merge remote-tracking branch 'upstream/main' into cache_docs

gante commented on docs/source/en/kv_cache.md:
```diff
@@ -0,0 +1,222 @@
+In 🤗 Transformers, we support various Cache types to optimize the performance across different models and tasks. By default all models generate with caching
+and for that we use the ["DynamicCache"] class. It allows us to dynamically grow cache size, by saving more and more keys and values as we generate. Usually all you have to do to enable or disable it is to pass in `use_cache=True` or `use_cache=False` into the `generate()` method.
+
+Apart from that we support many cache classes, which can be set with a `cache_implementation` argument. It can be set to one of `["static", "quantized", "sliding_window", "hybrid"]`. Let's look into each cache type in more details.
```
Instead of enumerating specific values for `cache_implementation` here, let's link to the field in the documentation (which should have the full list always up to date!)

gante commented on docs/source/en/kv_cache.md:
```diff
@@ -0,0 +1,222 @@
+
+### Quantized Cache
+
+The key and value cache can occupy a large portion of memory, becoming a bottleneck for long-context generation, especially for Large Language Models.
```
```suggestion
The key and value cache can occupy a large portion of memory, becoming a [bottleneck for long-context generation](https://huggingface.co/blog/llama31#inference-memory-requirements), especially for Large Language Models.
```
I think this link illustrates the problem very well!

gante commented on docs/source/en/kv_cache.md:
```diff
@@ -0,0 +1,222 @@
+
+<Tip warning={true}>
+
+Cache quantization can be detrimental if the context length is short and there is enough GPU VRAM available to run without cache quantization.
```
I know this is copied but... it's detrimental in what sense? Let's be more precise :D

gante commented on docs/source/en/kv_cache.md:
````diff
@@ -0,0 +1,222 @@
+>>> tokenizer.batch_decode(out, skip_special_tokens=True)[0]
+```
+
+### Sink Cache
````
Let's place this one last: 
1. sink cache is meant to solve a problem that is no longer critical (very short context lengths)
2. it often has bugs 😬

zucchini-nlp replied:
Oke, will add it. But I hope we're going to solve those bugs 😄  Seems like users still use it actively, not as much as StaticCache ofc

gante commented on docs/source/en/kv_cache.md:
```diff
@@ -0,0 +1,222 @@
+
+### Sink Cache
+
+Sink Cache was introduced in ["Efficient Streaming Language Models with Attention Sinks"](https://arxiv.org/abs/2309.17453). It allows you to generate long sequences of text ("infinite length" according to the paper) without any fine-tuning. That is achieved by smart handling of previous keys and values, specifically it retains a few initial tokens from the sequence, called "sink tokens". This is ...
```
```suggestion
Sink Cache was introduced in ["Efficient Streaming Language Models with Attention Sinks"](https://arxiv.org/abs/2309.17453). It allows you to generate long sequences of text ("infinite length" according to the paper) without any fine-tuning. That is achieved by smart handling of previous keys and values, specifically it retains a few initial tokens from the sequence, called "sink tokens". This is based on the observation that these initial tokens attract a significant portion of attention scores during the generation process. Tokens that come after "sink tokens" are discarded on a sliding windowed basis, keeping only the latest `window_size` tokens. By keeping these initial tokens as "attention sinks," the model maintains stable performance even when dealing with very long texts, thus discarding most of the previous knowledge.
```

gante commented on docs/source/en/kv_cache.md:
```diff
@@ -0,0 +1,222 @@
+
+>>> # simply pass the cache implementation="static"
+>>> out = model.generate(**inputs, do_sample=False, max_new_tokens=20, cache_implementation="static")
+>>> tokenizer.batch_decode(out, skip_special_tokens=True)[0]
```
missing expected output after this line (might be relevant to doctests)

👉  some other examples are missing it too!

gante commented on docs/source/en/kv_cache.md:
```diff
@@ -0,0 +1,222 @@
+
+Hybrid Cache is used in specific models only. It is a combination of SlidingWindow and Static Cache, where every layer in the model operates with one of these cache types. Currently only Gemma2 supports it and uses it be default whenever you load the model. Using Hybrid Cache with other models can results in performance degradation and unexpected generation results. It can be used by passing `cac ...
+
+
```
Missing section: ### Model-specific caches

We should mention that some models, like Jamba, are exclusively compatible with their own cache types, and that the model page often has more information about it (e.g. https://huggingface.co/docs/transformers/v4.43.2/en/model_doc/jamba#transformers.JambaForCausalLM.forward.past_key_values)

zucchini-nlp replied:
Ah, right, forgot about mamba and jamba. Gemma2 should be there too in that case

gante commented on docs/source/en/kv_cache.md:
````diff
@@ -0,0 +1,222 @@
+    messages.append({"role": "assistant", "content": completion})
+
+print(messages)
+```
````
plz add a TODO for us to add a section about cache reuse (which is buggy atm 🤗 )

ArthurZucker replied:
#32159 we can link this I'll test it soon

zucchini-nlp replied:
Wow looks interesting, like PrefixCache :)

gante commented on src/transformers/cache_utils.py:
```diff
@@ -296,6 +296,22 @@ class DynamicCache(Cache):
+        >>> # Prepare a cache class and pass it to model's forward
+        >>> past_key_values = DynamicCache()
+        >>> outputs = model(**inputs, past_key_values=past_key_values, use_cache=True)
+        >>> past_kv_length = outputs.past_key_values.get_seq_length()
```
missing on all examples: printing the generated output (it will be checked by our doctests 🤗 )

gante reviewed (commented):
💛 💛 💛 

Added a few improvement suggestions 🤗

zucchini-nlp commented on docs/source/en/kv_cache.md:
```diff
@@ -0,0 +1,222 @@
+including text generation, translation, summarization and other transformer-based applications.
+Effective caching helps reduce computation time and improve response rates, especially in real-time or resource-intensive applications.
+
+Transformers support various caching methods, leveraging `Cache` classes to abstract and manage the caching logic.
```
Ah, that's how we link, okey. I couldn't find a way to link to other docs, will fix

zucchini-nlp commented on docs/source/en/kv_cache.md:
```diff
@@ -0,0 +1,222 @@
+
+More concretely, key-value cache acts as a memory bank for these generative models, where the model stores key-value pairs derived from self-attention layers for previously processed tokens. By storing this information, the model can avoid redundant computations and instead retrieve keys and values of previous tokens from the cache.
+
+<details>
```
will try to follow the gist. For me it wouldn't render correctly with simpler syntax but maybe I was missing smth

gante replied:
make sure you have empty lines between the `<>` lines and correct indentation :)

zucchini-nlp committed:
Update docs/source/en/kv_cache.md

Co-authored-by: Joao Gante <joaofranciscocardosogante@gmail.com>

zucchini-nlp committed:
Update docs/source/en/kv_cache.md

Co-authored-by: Joao Gante <joaofranciscocardosogante@gmail.com>

zucchini-nlp committed:
Update docs/source/en/kv_cache.md

Co-authored-by: Joao Gante <joaofranciscocardosogante@gmail.com>

zucchini-nlp committed:
Update docs/source/en/kv_cache.md

Co-authored-by: Joao Gante <joaofranciscocardosogante@gmail.com>

zucchini-nlp committed:
Update docs/source/en/kv_cache.md

Co-authored-by: Joao Gante <joaofranciscocardosogante@gmail.com>

gante reviewed (commented):
One more thing! We should add throughout the text (in sereval places) that these classes are for decoder-only models, and, in its own subsection with an example, that we have a [wrapper class for encoder-decoder compatibility](https://github.com/huggingface/transformers/blob/b8e5cd5396f7c0cc2d5e10be6696ea38742abf51/src/transformers/cache_utils.py#L1000) :D

zucchini-nlp commented:
Sure, I will add a subsection for Encoder-decoder cache below with more details on how to use it and when. Ughf, how many cache classes we have added 🙃

ArthurZucker commented on docs/source/en/generation_strategies.md:
We can maybe mention / link to cache doc for KV Cache quantization. 
IMO we should also have a section in LLMOpitmisation or inference or both, linking to the kv_cache doc!

ArthurZucker commented on docs/source/en/kv_cache.md:
Super nice! One additional piece would be a table, with each cache function's name, and something like 

is memory efficient | is fast | supports compile | etc or even have a sample example with `llama3.1` and we show peak memory consumption for each cache class, time it took with compile

ArthurZucker commented on src/transformers/cache_utils.py:
```diff
@@ -1146,6 +1269,42 @@ def batch_select_indices(self, indices: torch.Tensor):
+        >>> max_generated_length = inputs.input_ids.shape[1] + 10
+        >>> past_key_values = HybridCache(config=model.config, max_batch_size=1, max_cache_len=max_generated_length, device=model.device, dtype=model.dtype)
+        >>> outputs = model(**inputs, past_key_values=past_key_values, use_cache=True)
+        >>> past_kv_length = outputs.past_key_values.get_seq_length()
```
I don't think we should show the get seq length but otherwise all good!

ArthurZucker reviewed (commented):
💘 very nice to add this, hope it will make discovery easier for everyone !

zucchini-nlp committed:
one more small update

ArthurZucker commented on src/transformers/cache_utils.py:
@gante this is making me think we could return the cache position in generate, this way you can continue generation without guessing where you are!

zucchini-nlp replied:
Btw why cache position cannot be an attr in the cache class if we will have same shape position after pre-fill? Will it break compile?

gante replied:
I would like very much for it to be a cache attribute (because it is, after all, a position of the cache), but our previous attempts at making it compileable failed 💔 

Working compilation > good object abstraction :D 

(Feel free to explore the topic, if it is of your interest! In any case, the most important part, i.e. having a working compilation, is solved 🤗 )

ArthurZucker reviewed (approved):
Very nice updates

gante commented on docs/source/en/kv_cache.md:
```diff
@@ -61,34 +67,46 @@ More concretely, key-value cache acts as a memory bank for these generative mode
+| Static Cache        |      No          |        Yes               |         Yes                |          |     No                   |
+| Quantized Cache     |      Yes         |        No                |         No                 |          |     Yes                  |
+| Sliding Window Cache|      No          |        Yes               |         Yes                |          |     No                   |
+| Hybrid Cache        |      No          |        Yes               |         Yes                |          |     No                   |
+| Sink Cache          |      Yes         |        No                |         Yes                |          |     Yes                  |
+| Mamba Cache         |      No          |        No                |         Yes                |          |     No                   |
```
Hybrid and Mamba shouldn't be here, since they are model-specific

gante commented on docs/source/en/kv_cache.md:
```diff
@@ -130,49 +148,142 @@ For more examples with Static Cache and JIT compilation, take a look at (StaticC
+Below, we provide detailed descriptions and usage guidelines for each specialized cache class:
+
+
 ### Hybrid Cache
```
Sorry, I wasn't clear in the previous review cycle :( 

I didn't mean to say we should have one subsection per model-specific cache, but rather that we should have a single section saying that some models have their own cache types (and are not compatible with the caches in this guide).

Having a subsection per model-specific cache will likely lead to bloated and quickly outdated docs :)

zucchini-nlp committed:
should render hidden secrtion now

zucchini-nlp committed:
add outputs

zucchini-nlp committed:
fix links

zucchini-nlp commented:
Okey, this is ready now. Did some final updates for the formatting and links, also addressed comments from @gante  and added a small paragraph about `cache_position` in the hidden section

I think we can add more on dialogue generation with `compile` and static cache later, prob it should go in the `llm-optims.md` file. I already asked Lewis (don't know his GH tag), who experimented with that, if there are any tricky parts we could shape with users. I can work on it in another PR

gante commented:
@zucchini-nlp Lewis -> @lewtun :)

gante commented on docs/source/en/kv_cache.md:
```diff
@@ -20,7 +20,7 @@ Efficient caching is crucial for optimizing the performance of models in various
 Effective caching helps reduce computation time and improve response rates, especially in real-time or resource-intensive applications.
 
-Transformers support various caching methods, leveraging [`~Cache`] classes to abstract and manage the caching logic.
+Transformers support various caching methods, leveraging "Cache" classes to abstract and manage the caching logic.
```
Are links like [\`~Cache\`] not working? 🤔 All public classes should be linkable through [\`~ClassName\`] 

(sorry to be persistent here, I'm spending more energy on these repo-wide automatic tools :) )

zucchini-nlp replied:
No, it didn't work for me, no idea why. Should I try again?

gante replied:
Yes please :) I'll check the CI job if it doesn't work 🙏

zucchini-nlp replied:
Sure thing! I will push the change soon and tag you when it fails :)

zucchini-nlp replied:
It works now magically, hehe. I think there was an error and I used another type of quote-marks/brackets 😄

gante reviewed (commented):
A final nit :o

zucchini-nlp committed:
check links

zucchini-nlp committed:
Merge branch 'main' into cache_docs

zucchini-nlp committed:
update all links

zucchini-nlp committed:
update with offloaded cache

zucchini-nlp committed:
all cache is importable, so they appear in docs

zucchini-nlp committed:
fix copies

zucchini-nlp committed:
docstring...

gante commented on src/transformers/cache_utils.py:
```diff
@@ -299,6 +299,22 @@ class DynamicCache(Cache):
+        >>> # Prepare a cache class and pass it to model's forward
+        >>> past_key_values = DynamicCache()
+        >>> outputs = model(**inputs, past_key_values=past_key_values, use_cache=True)
+        >>> past_kv_length = outputs.past_key_values # access cache filled with key/values from generation
```
Missing the output in these examples, so we can doctest :P

zucchini-nlp replied:
Not sure what should be the output here, as we are not printing anything. I wasn't very pro of getting kv values or logits slice, as it's prone to numerical precision errors

zucchini-nlp replied:
Yeah, I was afraid to cause more numerical precision fails, but can be added for the sake of testing

zucchini-nlp replied:
Yeah, I was afraid to cause more numerical precision fails, but can be added for the sake of testing

gante reviewed (approved):
Thank you iterating on the docs! 💛 

Left a minor optional nit, up to you:)
